File: Data Structures & Algorithms/valid-binary-search-tree/submission-5.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */


// so to satisfy the BST we have to ensure that at any given node
// the left child is less than the node
// the right child is greater than the node

// we'll need to pass in a range the child must satisfy

#include <climits>

class Solution 
{
public:
    bool isValidBST(TreeNode* root) 
    {
        return validate(root, LLONG_MIN, LLONG_MAX);
    }

    bool validate(TreeNode *root, long long low_bound, long long high_bound)
    {
        if (!root) { return true; }
        if (root->val <= low_bound || root->val >= high_bound) { return false; }

        return validate(root->left, low_bound, root->val) && validate(root->right, root->val, high_bound);
    }
};

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-5.py
# Lowest node T that both p and q are descendants.
# Initially p & q will be both on the left side or right side of cur node
# If p & q are both less than cur -> go left
# if p & q are both greater than cur -> go right
# otherwise this is the lowest common ancestor
    # beyond this point we'd only have at most p or q below us.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        # since p & q both exist we're safe to start searching
        # we'll never traverse all the way down, since we're guaranteed to find the ancestor
        cur = root

        while cur:
            if p.val < cur.val and q.val < cur.val:
                cur = cur.left
            elif p.val > cur.val and q.val > cur.val:
                cur = cur.right
            else:
                return cur


        